service/history/replicationTaskProcessor.go

```diff
@@ -0,0 +1,285 @@
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/service/worker/replicator"
+	"go.uber.org/yarpc/yarpcerrors"
+)
+
+const (
+	dropSyncShardTaskTimeThreshold = 10 * time.Minute
+	replicationTimeout             = 30 * time.Second
+)
+
+var (
+	// ErrUnknownReplicationTask is the error to indicate unknown replication task type
+	ErrUnknownReplicationTask = &shared.BadRequestError{Message: "unknown replication task"}
+)
+
+type (
+	replicationTaskProcessor struct {
+		shard            ShardContext
+		readLevel        int64
+		historyEngine    Engine
+		sourceCluster    string
+		domainReplicator replicator.DomainReplicator
+		metricsClient    metrics.Client
+		logger           log.Logger
+
+		requestChan chan<- *request
+		done        chan struct{}
+	}
+
+	request struct {
+		token    *r.ReplicationToken
+		respChan chan<- *r.ReplicationTasksInfo
+	}
+)
+
+func newReplicationTaskProcessor(
+	shard ShardContext,
+	historyEngine Engine,
+	domainReplicator replicator.DomainReplicator,
+	metricsClient metrics.Client,
+	replicationTaskFetcher *replicationTaskFetcher,
+) *replicationTaskProcessor {
+	return &replicationTaskProcessor{
+		shard:            shard,
+		historyEngine:    historyEngine,
+		sourceCluster:    replicationTaskFetcher.GetSourceCluster(),
+		domainReplicator: domainReplicator,
+		metricsClient:    metricsClient,
+		logger:           shard.GetLogger(),
+		requestChan:      replicationTaskFetcher.GetRequestChan(),
+		done:             make(chan struct{}),
+	}
+}
+
+func (p *replicationTaskProcessor) Start() {
+	go func() {
+		p.readLevel = p.shard.GetClusterReplicationLevel(p.sourceCluster)
+		scope := p.metricsClient.Scope(metrics.ReplicationTaskFetcherScope, metrics.TargetClusterTag(p.sourceCluster))
+
+		for {
+			respChan := make(chan *r.ReplicationTasksInfo, 1)
+			p.requestChan <- &request{
+				token:    &r.ReplicationToken{ShardID: common.Int32Ptr(int32(p.shard.GetShardID())), TaskID: common.Int64Ptr(p.readLevel)},
+				respChan: respChan,
+			}
+
+			select {
+			case response := <-respChan:
+				p.logger.Info("Got fetch replication tasks response.",
+					tag.ReadLevel(response.GetReadLevel()),
+					tag.Bool(response.GetHasMore()),
+					tag.Counter(len(response.GetReplicationTasks())),
+				)
+
+				for _, replicationTask := range response.ReplicationTasks {
+					p.processTask(replicationTask)
+				}
+
+				p.readLevel = response.GetReadLevel()
+				err := p.shard.UpdateClusterReplicationLevel(p.sourceCluster, p.readLevel)
+				if err != nil {
+					p.logger.Error("Error updating replication level for shard", tag.Error(err), tag.OperationFailed)
+				}
+
+				scope.UpdateGauge(metrics.ReplicationTasksReadLevel, float64(p.readLevel))
+				scope.AddCounter(metrics.ReplicationTasksApplied, int64(len(response.GetReplicationTasks())))
+			case <-p.done:
+				p.logger.Info("Closing replication task processor.", tag.ShardID(p.shard.GetShardID()), tag.ReadLevel(p.readLevel))
```

Comment:
This logger should already have ShardID because you get it from shardContext-- if not please fix it.

Reply:
ok cool.